src/auth/auth.service.ts

```
@@ -0,0 +1,157 @@
+import {
+  UnprocessableEntityException,
+  Injectable,
+  HttpException,
+  Inject,
+} from '@nestjs/common';
+import { IUsersService, USERS_SERVICE } from '../users/users.service';
+import { User } from '../users/interface/users.interface';
+import {
+  REFRESH_TOKENS_SERVICE,
+  IRefreshTokensService,
+} from '../refreshTokens/refreshTokens.service';
+import { RefreshToken } from '../refreshTokens/interface/refreshTokens.interface';
+import { TokenExpiredError } from 'jsonwebtoken';
+import { jwtConstants } from './constants';
+import { JwtService } from '@nestjs/jwt';
+import * as bcrypt from 'bcrypt';
+import {
+  TokenPayload,
+  RefreshTokenPayload,
+  TokenResponse,
+} from './interface/auth.interface';
+
+export const AUTH_SERVICE = 'AUTH SERVICE';
+export interface IAuthService {
+  validateUser(email: string, password: string): Promise<User>;
+  validateToken(token: string): Promise<void>;
+  generateAccessToken(user: User): Promise<string>;
+  generateRefreshToken(user: User, expiresIn: string): Promise<string>;
+  decodeRefreshToken(token: string): Promise<RefreshTokenPayload>;
+  getUserFromRefreshTokenPayload(payload: RefreshTokenPayload): Promise<User>;
+  resolveRefreshToken(encoded: string): Promise<User>;
+  createAccessTokenFromRefreshToken(refresh: string): Promise<string>;
+  login(user: User): Promise<TokenResponse>;
+}
+
+@Injectable()
+export class AuthService implements IAuthService {
+  constructor(
+    @Inject(REFRESH_TOKENS_SERVICE)
+    private readonly refreshTokensService: IRefreshTokensService,
+    @Inject(USERS_SERVICE) private readonly usersService: IUsersService,
+    private jwtService: JwtService,
+  ) {}
+
+  public async validateUser(email: string, password: string): Promise<User> {
+    const user: User = await this.usersService.findByEmail(email);
+    if (!user || !user.active) {
+      return null;
```

I'm not sure about `null`. Shouldn't we return the 401 error?

---

At least, "bad request" error return. Depends on how do we use this method

---

local-strategy guard calls this function to validate user by credentials, expects User instance(or undefined), should we return http error from here?